src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs

```diff
@@ -250,6 +256,79 @@ public AspNetCoreMvc2Integration(object actionDescriptor, object httpContext)
             }
         }
 
+        /// <summary>
+        /// Wrapper method used to catch unhandled exceptions in the incoming request pipeline for Microsoft.AspNetCore.Mvc.Core
+        /// </summary>
+        /// <param name="context">The DiagnosticSource that this extension method was called on.</param>
+        [InterceptMethod(
+            CallerAssembly = "Microsoft.AspNetCore.Mvc.Core",
+            TargetAssembly = "Microsoft.AspNetCore.Mvc.Core",
+            TargetType = ExceptionHookType)]
+        public static void Rethrow(object context)
+        {
+            AspNetCoreMvc2Integration integration = null;
+            const string methodName = nameof(Rethrow);
+
+            try
+            {
+                if (context.TryGetPropertyValue("HttpContext", out object httpContext))
+                {
+                    if (httpContext.TryGetPropertyValue("Items", out IDictionary<object, object> contextItems))
+                    {
+                        integration = contextItems?[HttpContextKey] as AspNetCoreMvc2Integration;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorExceptionForFilter($"Error accessing {nameof(AspNetCoreMvc2Integration)}.", ex);
+            }
+
+            Action<object> rethrow = null;
+
+            try
+            {
+                var paramTypes = new[]
+                {
+                    context.GetType()
+                };
+
+                var methodKey = RethrowCache.GetMethodKey(paramTypes);
```

Comment:
I think I can reduce all the below code to a single method call on the RethrowCache object, which should greatly reduce the boiler plate we have in our integrations.
Anywhere this cache object is used would make the wrappers we swap in more robust, and less prone to type exceptions.

Reply:
Though performance needs to be tested.

Reply:
If this is boilerplate, let's do it now without changing other integrations. Then let's schedule a follow-up work to propagate the change everywhere.